### PR TITLE
Cleanup object states

### PR DESCRIPTION
**Pull request by cremebrule**

### CODE REVIEW

File: omnigibson/object_states/heat_source_or_sink.py [hunk 1]
```diff
@@ -81,6 +81,16 @@ def __init__(
...
         self.requires_inside = requires_inside
 
+    @classproperty
+    def metalink_prefix(cls):
+        return m.HEATSOURCE_LINK_PREFIX
+
+    @property
+    def _default_link(self):
+        # Only supported if we require inside
+        # return self.obj.root_link if self.requires_inside else super()._default_link
```
Comment (ChengshuLi) [inline omnigibson/object_states/heat_source_or_sink.py:91]: remove

File: omnigibson/object_states/link_based_state_mixin.py [hunk 2]
```diff
@@ -9,31 +10,49 @@ class LinkBasedStateMixin:
...
+    @property
+    def _default_link(self):
+        """
+        Returns:
+            None or RigidPrim: If supported, the fallback link associated with this link-based state if
+                no valid metalink is found
+        """
+        # No default link by default
+        # TODO: Make this None to enforce that downstream object states have properly annotated objects
+        return self.obj.root_link
```
Comment (ChengshuLi) [inline omnigibson/object_states/link_based_state_mixin.py:40]: make it None?

File: omnigibson/object_states/link_based_state_mixin.py [hunk 3]
```diff
@@ -9,31 +10,49 @@ class LinkBasedStateMixin:
...
+        # default_link to use
         if self.link is None:
-            return None
-
-        pos, _ = self.link.get_position_orientation()
-        return np.array(pos)
\ No newline at end of file
+            raise ValueError(f"Error: failed to initialize LinkBasedStateMixin {self.__class__.__name__} for object "
+                             f"{self.obj.name}; no metalink with prefix {self.metalink_prefix} found!")
+        return True
```
Comment (ChengshuLi) [inline omnigibson/object_states/link_based_state_mixin.py:58]: No need to return True if it never returns False?

File: omnigibson/object_states/particle_modifier.py [hunk 4]
```diff
@@ -323,7 +328,7 @@ def check_overlap():
             # Define the function for checking whether points are within the adjacency mesh
             def check_in_adjacency_mesh(particle_positions):
                 # Define the AABB bounds
-                lower, upper = self.obj.states[AABB].get_value() if self.link is None else self.link.aabb
+                lower, upper = self.link.aabb
```
Comment (ChengshuLi) [inline omnigibson/object_states/particle_modifier.py:331]: hmmm when `self.link is None`, shouldn't we still use the object bounding box instead of base link bounding box?

File: omnigibson/object_states/particle_modifier.py [hunk 5]
```diff
@@ -378,6 +383,7 @@ def _generate_limit_condition(self, system):
             function: Limit checker function, with signature condition(obj) --> bool, where @obj is the specific object
                 that this ParticleModifier state belongs to
         """
+        # TODO: Update to only check for systems within conditions
```
Comment (ChengshuLi) [inline omnigibson/object_states/particle_modifier.py:386]: I already did this in my PR.

File: omnigibson/object_states/particle_modifier.py [hunk 6]
```diff
@@ -624,10 +635,45 @@ def physical_particle_modification_limit(self):
...
+            for this particle modifier. Expected entries are as follows:
+
+                "type": (str), one of {"Cylinder", "Cone", "Sphere"}
+                "extents": (3-array), the (x,y,z) extents of the generated volume (specified in local link frame!)
+                "visualize": (bool), whether to visualize this projection or not
+
+            If None, information found from @obj.metadata will be used instead.
+        sample_with_raycast (bool): If True, will only sample particles at raycast hits. Otherwise, will bypass sampling
+            and immediately sample particles at the sampled particle locations. Note that this will only work
+            for PhysicalParticleSystem-based ParticleAppliers that use the Projection method!
```
Comment (ChengshuLi) [inline omnigibson/object_states/particle_modifier.py:665]: Sorry what is the use case for this exactly?

File: omnigibson/object_states/particle_modifier.py [hunk 7]
```diff
@@ -723,7 +823,53 @@ def _apply_particles_at_raycast_hits(self, system, hits):
             n_particles = min(len(hits), m.PHYSICAL_PARTICLES_APPLICATION_LIMIT - self.modified_particle_count[system])
             # Generate particles
             if n_particles > 0:
-                system.default_particle_instancer.add_particles(positions=np.array([hit[0] for hit in hits[:n_particles]]))
+                velocities = None if self._initial_speed == 0 else -self._initial_speed * np.array([hit[1] for hit in hits[:n_particles]])
```
Comment (ChengshuLi) [inline omnigibson/object_states/particle_modifier.py:826]: hmm why do we want particles to have initial speed?

File: omnigibson/object_states/particle_modifier.py [hunk 8]
```diff
@@ -30,8 +30,8 @@
 # Create settings for this module
 m = create_module_macros(module_path=__file__)
 
-m.APPLICATION_LINK_NAME = "particleapplication_link"
-m.REMOVAL_LINK_NAME = "particleremover_link"
+m.APPLICATION_LINK_PREFIX = "particleapplication"
```
Comment (ChengshuLi) [inline omnigibson/object_states/particle_modifier.py:33]: Note for myself, need to review this file more carefully once again.

File: omnigibson/prims/rigid_prim.py [hunk 9]
```diff
@@ -101,8 +101,8 @@ def _post_load(self):
         self.update_meshes()
 
         # Possibly set the mass / density
-        if len(self._collision_meshes) == 0:
-            # We have no collision meshes, so set a negligible mass
+        if len(self._collision_meshes) == 0 and len(self._visual_meshes) > 0:
```
Comment (ChengshuLi) [inline omnigibson/prims/rigid_prim.py:104]: when `len(self._collision_meshes) == 0 and len(self._visual_meshes) == 0)`, what should we do?

Comment (ChengshuLi) [reply omnigibson/prims/rigid_prim.py:104]: in any case, let's go over all the special casing for mass/density together to make sure it's rock solid.

File: omnigibson/objects/object_base.py [hunk 10]
```diff
@@ -221,7 +223,7 @@ def articulation_root_path(self):
         if self.kinematic_only or ((not has_articulated_joints) and (not has_fixed_joints)):
             # Kinematic only, or non-jointed single body objects
             return None
-        elif has_articulated_joints and not self.fixed_base:
+        elif (has_articulated_joints or has_fixed_joints) and not self.fixed_base:
```
Comment (ChengshuLi) [inline omnigibson/objects/object_base.py:226]: hmm the current version is still buggy?

File: omnigibson/object_states/temperature.py [hunk 11]
```diff
@@ -93,6 +74,9 @@ def _update(self):
 
         self.value = new_temperature
 
+        # Also update cache to force syncing of temperature so get_value() is consistent with self.value
+        self.update_cache(get_value_args=())
```
Comment (ChengshuLi) [inline omnigibson/object_states/temperature.py:78]: hmm why is this needed? seems fairly hacky.

**Review by ChengshuLi — COMMENTED** [review]
I review part of this PR. Will keep reviewing it a bit later.

File: omnigibson/object_states/particle_modifier.py [hunk 12]
```diff
@@ -826,9 +974,14 @@ def _get_max_particles_limit_per_step(self, system):
...
-    def get_state_link_name():
-        return m.APPLICATION_LINK_NAME
+    @classproperty
+    def metalink_prefix(cls):
+        return m.APPLICATION_LINK_PREFIX
+
+    @property
+    def _default_link(self):
+        # Only supported for adjacency, NOT projection
+        return self.obj.root_link if self.method == ParticleModifyMethod.ADJACENCY else super()._default_link
```
Comment (ChengshuLi) [inline omnigibson/object_states/particle_modifier.py:984]: Same here
```
return self.obj.root_link if self.method == ParticleModifyMethod.ADJACENCY else None
```

File: omnigibson/prims/rigid_prim.py (hunk 9 above)
Comment (ChengshuLi) [reply omnigibson/prims/rigid_prim.py:104]: `if len(self._collision_meshes) == 0`, don't apply mass API?

File: omnigibson/object_states/temperature.py (hunk 11 above)
Comment (ChengshuLi) [reply omnigibson/object_states/temperature.py:78]: Remove this. Also remove circular import in HeatSourceOrSink, and re-implement `compute_temperature_delta(self, obj)` as `affect_object(self, obj)`, which return True/False.

File: omnigibson/objects/primitive_object.py [hunk 13]
```diff
@@ -186,6 +176,21 @@ def _post_load(self):
...
+        # Set color and opacity
+        if self._prim_type == PrimType.RIGID:
+            visual_geom_prim = list(self.links["base_link"].visual_meshes.values())[0]
+        elif self._prim_type == PrimType.CLOTH:
+            visual_geom_prim = self.links["base_link"]
+        else:
+            raise ValueError("Prim type must either be PrimType.RIGID or PrimType.CLOTH for loading a primitive object")
+
+        visual_geom_prim.color = self._load_config["color"]
+        visual_geom_prim.opacity = self._load_config["opacity"]
```
[the remark this answers is not in the document]
Comment (ChengshuLi) [reply omnigibson/objects/primitive_object.py:192]: batch material.force_update

File: omnigibson/object_states/attachment.py [hunk 14]
```diff
@@ -150,6 +150,11 @@ def _attach(self, other):
         attached_joint.GetAttribute("physics:localPos1").Set(Gf.Vec3f(0.0, 0.0, 0.0))
         attached_joint.GetAttribute("physics:localRot1").Set(Gf.Quatf(1.0, 0.0, 0.0, 0.0))
 
+        # We update the simulation now without actually stepping physics so we can bypass the snapping warning from
+        # PhysicsUSD
+        with suppress_omni_log(channels=["omni.physx.plugin"]):
+            og.sim.pi.update_simulation(elapsedStep=0, currentTime=og.sim.current_time)
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/attachment.py:156]: done!

File: omnigibson/object_states/heat_source_or_sink.py (hunk 1 above)
Comment (cremebrule) [reply omnigibson/object_states/heat_source_or_sink.py:91]: done!

File: omnigibson/object_states/link_based_state_mixin.py (hunk 2 above)
Comment (cremebrule) [reply omnigibson/object_states/link_based_state_mixin.py:40]: updated the todo!

File: omnigibson/object_states/link_based_state_mixin.py (hunk 3 above)
Comment (cremebrule) [reply omnigibson/object_states/link_based_state_mixin.py:58]: done!

File: omnigibson/object_states/particle_modifier.py (hunk 4 above)
Comment (cremebrule) [reply omnigibson/object_states/particle_modifier.py:331]: keep as is, per our discussion

File: omnigibson/object_states/particle_modifier.py (hunk 5 above)
Comment (cremebrule) [reply omnigibson/object_states/particle_modifier.py:386]: removed!

File: omnigibson/simulator.py [hunk 15]
```diff
@@ -403,6 +406,12 @@ def _transition_rule_step(self):
         """
         Applies all internal non-Omniverse transition rules.
         """
+        # Apply any transiiton object init states from before, and then clear the dictionary
+        for obj, states_info in self._transition_object_init_states.items():
+            for state, args in states_info.items():
+                obj.states[state].set_value(*args)
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/simulator.py:412]: yep!

File: omnigibson/objects/light_object.py [hunk 16]
```diff
@@ -120,6 +120,9 @@ def _post_load(self):
         # run super first
         super()._post_load()
 
+        # Make root link mass very small so omni doesn't complain about no mass
+        self.root_link.mass = 1e-6
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/objects/light_object.py:124]: in general, todo for future PR, but removed this line because it's already taken care of in RigidPrim!

File: omnigibson/simulator.py [hunk 17]
```diff
@@ -590,22 +606,31 @@ def _on_contact(self, contact_headers, contact_data):
...
+                actor1_obj = self._scene.object_registry("prim_path", "/".join(actor1.split("/")[:-1]))
+                if actor0_obj is None or actor1_obj is None or not actor0_obj.initialized or not actor1_obj.initialized:
                     continue
-                for state_type in obj0.states:
-                    if not issubclass(state_type, ContactSubscribedStateMixin):
+                elif (actor1_obj, actor0_obj) in combos:
+                    headers[(actor1_obj, actor0_obj)].append(contact_header)
+                else:
+                    combos.add((actor0_obj, actor1_obj))
+                    headers[(actor0_obj, actor1_obj)].append(contact_header)
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/simulator.py:624]: done!

File: omnigibson/object_states/particle_source_or_sink.py [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+
+        # Convert inputs into arguments to pass to particle applier class
+        super().__init__(
+            obj=obj,
+            method=ParticleModifyMethod.PROJECTION,
+            conditions=conditions,
+            projection_mesh_params={
+                "type": "Cylinder",
+                "extents": [sink_radius * 2, sink_radius * 2, sink_height],
+                "visualize": False,
```
[the remark this answers is not in the document]
Comment (cremebrule) [reply omnigibson/object_states/particle_source_or_sink.py:143]: TODO: Discuss further w/ Cem / Arman

File: omnigibson/object_states/particle_source_or_sink.py [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+            },
+        )
+
+    def _initialize(self):
+        # Run super first
+        super()._initialize()
+
+        # Override check overlap such that it always returns True (since we are ignoring overlaps and directly
+        # removing particles
+        self._check_overlap = lambda: True
```
Comment (ChengshuLi) [inline omnigibson/object_states/particle_source_or_sink.py:150]: Sorry why are we ignoring overlaps and directly removing particles?

File: omnigibson/object_states/temperature.py [hunk 20]
```diff
@@ -60,30 +61,10 @@ def _update(self):
...
-                        continue
-
-                new_temperature += (
-                    (heat_source.temperature - self.value) * heat_source.heating_rate * self._simulator.get_rendering_dt()
-                )
-                affected_by_heat_source = True
+
+            # Compute delta to apply if the heat source is actively affecting this object
+            if heat_source.affects_obj(obj=self.obj):
+                new_temperature += (heat_source.temperature - self.value) * heat_source.heating_rate * og.sim.get_rendering_dt()
```
Comment (ChengshuLi) [inline omnigibson/object_states/temperature.py:67]: should set `affected_by_heat_source` as True here?

File: omnigibson/objects/controllable_object.py [hunk 21]
```diff
@@ -343,7 +347,7 @@ def _actions_to_control(self, action):
         # Compose controls
         u_vec = np.zeros(self.n_dof)
         # By default, the control type is effort and the control value is 0 (np.zeros) - 0 effort means no control.
-        u_type_vec = np.array([ControlType.EFFORT] * self.n_dof)
+        u_type_vec = np.array([ControlType.NONE] * self.n_dof)
```
Comment (ChengshuLi) [inline omnigibson/objects/controllable_object.py:350]: Update the comment above?

File: omnigibson/prims/entity_prim.py [hunk 22]
```diff
@@ -375,6 +375,7 @@ def materials(self):
                     mat_path = xform.material.prim_path
                     if mat_path not in material_paths:
                         materials.add(xform.material)
+                        material_paths.add(mat_path)
```
Comment (ChengshuLi) [inline omnigibson/prims/entity_prim.py:378]: wow I can't believe I missed this. Did this cause some concrete bugs?

File: omnigibson/prims/joint_prim.py [hunk 23]
```diff
@@ -793,8 +805,9 @@ def set_effort(self, effort, normalized=False):
                 de-normalized first before being executed). Default is False
         """
         # Sanity checks -- make sure that we're articulated (no control type check like position and velocity
-        # because we can't set effort targets)
+        # because we can't set effort targets) and that we're driven
         self.assert_articulated()
+        assert self._driven, "Cannot set efforts for joint that is not driven!"
```
Comment (ChengshuLi) [inline omnigibson/prims/joint_prim.py:810]: If it's not driven, can you set_vel and set_pos? Let's maybe discuss in person.

File: omnigibson/prims/rigid_prim.py [hunk 24]
```diff
@@ -476,6 +476,11 @@ def mass(self, mass):
             mass (float): mass of the rigid body in kg.
         """
         self._mass_api.GetMassAttr().Set(mass)
+        # Write to dc as well if it's accessible
+        if self.dc_is_accessible:
+            rb_props = self.rigid_body_properties
+            rb_props.mass = mass
+            self._dc.set_rigid_body_properties(self._handle, rb_props)
```
Comment (ChengshuLi) [inline omnigibson/prims/rigid_prim.py:483]: hmmm what is the relationship between these two different mass APIs? Also, does dc has a density API too that we need to incorporate?

File: omnigibson/robots/tiago.py [hunk 25]
```diff
@@ -282,7 +280,7 @@ def _post_load(self):
             self.eef_links[arm].visual_only = True
             self.eef_links[arm].visible = False
 
-        self._world_base_fixed_joint_prim = get_prim_at_path(os.path.join(self.root_link.prim_path, "world_base_joint"))
+        self._world_base_fixed_joint_prim = get_prim_at_path(f"{self._prim_path}/rootJoint")
```
Comment (ChengshuLi) [inline omnigibson/robots/tiago.py:283]: Did you also remove the fixed joint from Tiago in the `omni` branch of `ig_assets`?

File: omnigibson/scenes/scene_base.py [hunk 26]
```diff
@@ -4,19 +4,26 @@
...
+from pxr import Sdf, Gf
 
 # Create module logger
 log = create_module_logger(module_name=__name__)
 
+# Create settings for this module
+m = create_module_macros(module_path=__file__)
+
+# Default texture to use for skybox
+m.DEFAULT_SKYBOX_TEXTURE = f"{og.assets_path}/models/background/sky.jpg"
```
Comment (ChengshuLi) [inline omnigibson/scenes/scene_base.py:25]: same here, assets updated?

**Review by ChengshuLi — APPROVED** [review]
LGTM with minor comments.

File: omnigibson/object_states/temperature.py (hunk 20 above)
Comment (cremebrule) [reply omnigibson/object_states/temperature.py:67]: done!

File: omnigibson/objects/controllable_object.py (hunk 21 above)
Comment (cremebrule) [reply omnigibson/objects/controllable_object.py:350]: done!

File: omnigibson/prims/joint_prim.py (hunk 23 above)
Comment (cremebrule) [reply omnigibson/prims/joint_prim.py:814]: changed `target` to `drive` for better clarity

File: omnigibson/prims/rigid_prim.py (hunk 24 above)
Comment (cremebrule) [reply omnigibson/prims/rigid_prim.py]: turned out to be redundant, removed
